### PR TITLE
Create Actions and Transitions in C#

### LINKED ISSUE #86: Handle `ObjectChage` Action Type

### LINKED ISSUE #106: Create Action for Transition.Movement

### LINKED ISSUE #107: Create Action for Transition.MoveRelative

### LINKED ISSUE #108: Create Action for Transition.Color

### LINKED ISSUE #109: Create Action for Transition.Scale

### LINKED ISSUE #110: Create Action for Transition.Sound
Sound just plays when button is clicked? Not sure about the actual outcome

### LINKED ISSUE #111: Create Action for Transition.Link

### PR DESCRIPTION
**Pull request by RobertGemmaJr**
There's a fair bit of refactoring I want to do here but the PR is already pretty massive so I want to get eyes on it now. I'll probably review the code myself and add some discussion questions but here's what the PR accomplishes:

- Fixes catching and throwing of errors in `CLI.cs`
- Removes all UnityEngine code from `Classes.cs`
- `AddAction` and `GetTransition` helper function
- Updates the object and canvas prefabs with the correct tag
- Adds the `onClick` functions inside of `ButtonManager.cs`
- `ButtonManager.cs` script gets added to all of the objects
  - Contains all of the `onClick` actions for an object
- Adds the `Writing3D.Transitions` and `Writing3D.Actions` namespaces
- Handles the ObjectAction as `ScriptableObject`s in `Scripts/Actions`
- Handles the Transitions as `ScriptableObject`s in `Scripts/Transitions`

Each `<Link>` object in the XML contains a list of `<Actions>`. These actions can be one of several types and some  types contain a `<Transition>`, which in and of itself can be one of several types. These determine what occurs when a button is clicked in the project - making an `<Object>` invisible, moving it, changing color, playing a sound, etc etc etc. This gets into Unity's event system.

## Unity Event System

`UnityEvent`s and `UnityAction`s are a whole wormhole I've gone down and up for the last few weeks so I'll try to be detailed here. [button].onClick is an _event_ on which you can have many _actions_. The "normal" way is to just call `onClick.AddListener` at runtime (Unity ships with functions like Start() and Update(), more [here](https://docs.unity3d.com/Manual/ExecutionOrder.html)) just like you add callbacks in web development. But it doesn't make sense for me to be generating C# code (even if it is possible) and it would mean future students would have to write their own scripts which it definitely going to be too much for a writing class. Instead I'm utilizing `UnityEventTools.AddPersistentListener`. These can be generated in edit mode, plus they show up nicely in the GUI.

image
image

"Runtime only" indicates the callbacks will only fire when playing the game/project and there are otherwise 3 main parts. The bottom left is the GameObject the event is going to talk to. In the above example they all are referencing the button itself <sup>[1]</sup> but it can be any GameObject in the scene - more on that later. The top right is the function to be called and (when applicable) the bottom right is the argument that function is called with. So the above onClick's execution is:

1) `Find the [button] GameObject` -> `Get the  ButtonManager component (class instance) attached to it` -> `Call its Counter function`
    - This increases the `clickCount` private variable of the ButtonManager instance
2) `Find the [button] GameObject` -> `Get the  ButtonManager component (class instance) attached to it` -> `Call its ExecuteAction function with the [Link Action] data`
3) Repeat with a different [LinkAction]
4) Repeat with a different [LinkAction]
5) Repeat with a different [LinkAction]
6) `Find the [button] GameObject` -> `Get the  ButtonManager component (class instance) attached to it` -> `Call its Disable function`
    - This turns the `Button` component of the [button] GameObject's interactable state to false

<sup>1</sup> The `onClick` event is a member of the button component NOT a GameObject. GameObjects in Unity are really placeholders, all of their functionality comes from the components that are attached to them. So this would more accurately be written as "they are all referencing the GameObject they are attached to" 

## ScriptableObjects 

The major disadvantage of the persistent event listener is that the actions must be void or take exactly 1 generic type - see [here](https://docs.unity3d.com/ScriptReference/Events.UnityEventTools.html) in the docs. Moreover, the `AddObjectPersistantListener` must take ScriptableObject. A class instance is stored in memory which is re-initialized (to null) at runtime.

I honestly think the best way to think about a scriptable object is to think of it more like an xml or json file. It is a savable piece of data that is read at runtime by Unity. The major benefit is that you can attach functions to them and (much more importantly) Unity saves them as part of the components. This means they show up nicely in the editor and are NOT their own file. If every scriptable object needed to be saved in a file it would get very complicated very quickly. You can also create custom GUIs for them which is definitely something I'm looking to do in a separate PR, although it's definitely a lower priority.

## The Setup

Here's the xml associated with the second action from above:
```xml
<Actions>
    <ObjectChange name="frontlink">
        <Transition duration="1.0">
            <Visible>false</Visible>
        </Transition>
    </ObjectChange>
</Actions>
```

Note that it is missing a key optional piece - `<Clicks>`. It denotes how many times the button has to be clicked before the action is fired which defaults to 1 here. `[action].Invoke()` always executes every callback function inside of it, which is why I can't add the actions directly to `button.OnClick` - I have no way to check against NumClicks. Hence the `LinkAction` object and why `button.onClick` always calls `ButtonManager.ExecuteAction`. Each `LinkAction` then containts a custom `ActionEvent` (just an overlay for a `UnityEvent`, note the name in the screenshot below), which will always contain exactly 1 action<sup>1</sup>. It's annoying bloat but there's no way around it unfortunately.

image

Here you can see the `Clicks` default and that the action has been added to the event with a nested scriptable object. The callback is the `VisibleTransition` function of the `ObjectManager` class instance attached to the `movingtext` GameObject.

It's important to note here that `<ObjectAction>` is one of several different types of `<Action>`s and `<Movement>` is one of several different types of Transitions. It's all I have done so far but the concept will remain the same, only different data will be stored in the different types of ScriptableObjects. And for completions sake. here's what `Visible` looks like:

image

<sup>1</sup> At least as far as the archival projects are concerned, you certainly could add multiple in the future.

## Closing

Phew, okay, that was a lot. It's certainly been a lot of learning for me. Like I said I would really like to refactor this some more. Based on the other action types I'm not sure if I should go a step further with the abstraction - making the GUI cleaner but I wouldn't be able to use the separate switches I have going in `Helper.cs` and `Transition.cs` (The typing for UnityAction is very strict) or if I should go a step deeper and add a TransitionEvent to the Transition SO class. This would make the callback functions as simple as possible (e.g. VisibleTransition would just take a boolean). Will have to continue to experiment to see what's best

closes #86 
closes #106 
closes #107 
closes #108 
closes #109 
closes #110 
closes #111

### CODE REVIEW

**Comment by RobertGemmaJr** [comment]
Note that this does include changes from #117

**Review by kmilo9999 — COMMENTED** [review]
It took me a while to follow the logic and components you are using so far. I have a few notes here:
1. Do the `ObjectManager` and `ButtonManager` classes need to derived from `MonoBehaviour` ? I dont see it using any of the methods inherit from the base class.
2. Add the [gitignore ](https://github.com/github/gitignore/blob/main/Unity.gitignore) file recommended for unity projects.
3. For the `Helper.cs`. It might worth to check the [builder pattern](https://refactoring.guru/design-patterns/builder/csharp/example). We can address it in another PR.

**Comment by RobertGemmaJr** [comment]
> 1. Do the `ObjectManager` and `ButtonManager` classes need to derived from `MonoBehaviour` ? I dont see it using any of the methods inherit from the base class.

Yes, any script that's operating at runtime should derive from MonoBehavior. I'm not using anything right now but it will, especially in regards to interacting with other GameObjects in the scene.

> 2. Add the [gitignore ](https://github.com/github/gitignore/blob/main/Unity.gitignore) file recommended for unity projects.

It's in there. There's a second gitignore that's the unity default inside the [CAVE subfolder](https://github.com/brown-ccv/w3d-translator/blob/main/unity/CAVE/.gitignore)

> 3. For the `Helper.cs`. It might worth to check the [builder pattern](https://refactoring.guru/design-patterns/builder/csharp/example). We can address it in another PR.

This is awesome! Definitely another PR but I think this well help a lot. Thanks for the pointer Camilo!

**Comment by RobertGemmaJr** [comment]
I ended up merging the refactor right into this branch. I like how things are set up for now, although a later refactor may be in order once I start adding the different action typea

**Review by kmilo9999 — COMMENTED** [review]
It looks highly complex at this moment. Two things though:

1. About the Monobehaviour on `ButtonManager`. I see it being used for passing callback functions (Actions) to `AddVoidPersistentListener`. I know it's a script component that you are adding to the buttons. But I wonder (Or correct me if I am not following properly) if there's is way to decouple the callbacks code from the game objects method calls. 
2. The inheritance in Transition is all public. Do you need to encapsulate some of the members variables of the base class?

**Comment by RobertGemmaJr** [comment]
> 1. ... a way to decouple the callbacks code from the game objects method calls.

Yeah, that was a design decision I made, both ways are possible. Either every `<Object>` type gets an `ObjectManager` script (Group gets GroupManager, Sound gets SoundManager, etc) or every `Action` scriptable objects has a `Reference` property that contains the GameObject to be changed.

I think the way it's currently set up is more intuitive from the GUI and generally seems to be how to take advantage of Unity's event system. You can see which GameObject the action was associated with. If we were to add the reference to the `Action` (`Move` in the below screenshot) than the script reference in the bottom left would always be the same  GameObject (e.g. I would attach a single script instance to some root gameobject).

The scripts are small and we're not using the `Update` function so the performance hit is negligible. 

image

> 2. The inheritance in Transition is all public. Do you need to encapsulate some of the members variables of the base class?

I know it's not the safest way but I think it's okay to leave it as it is. There is no such thing as the base transition in the xml - it's just that every transition type has a `Duration` property. Having the base class makes it much easier to pass the variables between functions (without using `dynamic` and losing the IDE code completion). It's a bit less typing to have it as a member of the base class even though I always reference it from the derived class.

File: unity/CAVE/Assets/Editor/CLI.cs [hunk 1]
```diff
@@ -3,29 +3,25 @@
...
 // TODO (80): Should ConvertVector3 invert z axis always?
 
-// TODO: Naming scheme (prepend Xml)
-
 namespace Writing3D
 {
     namespace Translation
     {
-        public static class CLI
+        public static partial class CLI
```
Comment (eldu) [inline unity/CAVE/Assets/Editor/CLI.cs:24]: Why are we making this a partial class?

**Comment by RobertGemmaJr** [comment]
I'm going to merge this to get the ball rolling, I have a fairly large refactor in the pipeline

**Pull request merged**